### PR TITLE
feat: custom conflict resolution implementation in ObjC

### PR DESCRIPTION
**Pull request by jayahariv**
* adds interface for ConflictResolver, ConflictResolution
* adds the default conflict resolution static function.
* adds property to the replication configuration instance. not handled the property for now.
* add Conflict class with internal constructor and readOnly properties exposed. 
* implemented the conflict resolution case.
* exception handling add

Ref:  #2427

### CODE REVIEW

File: Objective-C/CBLConflictResolution.m [hunk 1]
```diff
@@ -0,0 +1,40 @@
+//
+//  CBLConflictResolution.m
+//  CBL ObjC
+//
+//  Created by Jayahari Vavachan on 4/26/19.
+//  Copyright © 2019 Couchbase. All rights reserved.
```
Comment (jayahariv) [inline Objective-C/CBLConflictResolution.m]: will change this!

Comment (jayahariv) [reply Objective-C/CBLConflictResolution.m]: done!!

**Comment by jayahariv** [comment]
taking this back, since the Conflict class seems to be declared separately. so that we can define the unavailable init, and make the implementation internal.

File: Objective-C/CBLConflict.m [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+@implementation CBLConflict
+
+@synthesize localDocument=_localDocument, remoteDocument=_remoteDocument;
+
+# pragma mark - Internal
+
+- (instancetype) initWithLocalDocument:(CBLDocument *)localDoc
+                        remoteDocument:(CBLDocument *)remoteDoc {
+    NSParameterAssert(localDoc != nil && remoteDoc != nil);
```
Comment (jayahariv) [inline Objective-C/CBLConflict.m]: it is nullable!!

Comment (jayahariv) [reply Objective-C/CBLConflict.m]:
```suggestion
```

File: Objective-C/CBLConflict.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
+//
+//  CBLConflict.h
+//  CouchbaseLite
+//
+//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
```
Comment (pasin) [inline Objective-C/CBLConflict.h]: should be 2019.

File: Objective-C/CBLConflictResolution.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+//
+//  CBLConflictResolution.h
+//  CouchbaseLite
+//
+//  Copyright (c) 2018 Couchbase, Inc All rights reserved.
```
Comment (pasin) [inline Objective-C/CBLConflictResolution.h]: Should be 2019

File: Objective-C/CBLConflictResolver.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
+//
+//  CBLConflictResolver.h
+//  CouchbaseLite
+//
+//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
```
Comment (pasin) [inline Objective-C/CBLConflictResolver.h]: 2019

File: Objective-C/CBLDatabase.mm [hunk 6]
```diff
@@ -950,102 +952,105 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
         
         // Resolve conflict:
         CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
                    docID, localDoc.revID, remoteDoc.revID);
+        CBLDocument* resolvedDoc = [conflictResolver resolve: conflict];
         
-        CBLDocument* resolvedDoc = [self resolveConflictBetweenLocalDoc: localDoc
-                                                           andRemoteDoc: remoteDoc];
+        if (resolvedDoc == nil) {
+            resolvedDoc = remoteDoc.isDeleted ? remoteDoc : localDoc;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: It's possible that both remoteDoc and localDoc are not deleted doc. There might not be a real use case but the API allows it. I think you could just pass null to saveResolvedDocument() then saveResolvedDocument() will determine it as a deleted doc.

File: Objective-C/CBLDatabase.mm [hunk 7]
```diff
@@ -950,102 +952,105 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
-        // Read the conflicting remote revision:
-        CBLDocument* remoteDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                            documentID: docID
-                                                        includeDeleted: YES
-                                                                 error: outError];
-        if (!remoteDoc || ![remoteDoc selectConflictingRevision])
-            return false;
+        conflictResolver = conflictResolver ?: [CBLConflictResolution default];
+        CBLConflict* conflict = [[CBLConflict alloc] initWithLocalDocument: localDoc
+                                                            remoteDocument: remoteDoc];
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: Need to check whether localDoc or remoteDoc is deleted doc or not. If it's a deleted doc, need to pass null. You may also add an assertion to make sure that both of them are not null.

File: Objective-C/CBLReplicator.mm [hunk 8]
```diff
@@ -73,6 +73,7 @@ @interface CBLReplicator ()
 @implementation CBLReplicator
 {
     dispatch_queue_t _dispatchQueue;
+    dispatch_queue_t _otherDispatchQueue; // TODO: Naming???
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: How about _conflictResolverDispatchQueue?

File: Objective-C/Internal/Replicator/CBLConflict+Internal.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
+//
+//  CBLConflict+Internal.h
+//  CBL ObjC
+//
+//  Created by Jayahari Vavachan on 4/26/19.
```
Comment (pasin) [inline Objective-C/Internal/Replicator/CBLConflict+Internal.h]: Fix header.

File: Objective-C/CBLReplicator.mm [hunk 10]
```diff
@@ -551,13 +556,18 @@ - (void) onDocsEnded: (NSArray<CBLReplicatedDocument*>*)docs pushing: (BOOL)push
...
-            NSError* error;
-            if ([_config.database resolveConflictInDocument: doc.id error: &error]) {
-                [doc resetError]; // Reset the error as successfully resolving the conflicts
-            } else
-                CBLWarn(Sync, @"%@: Conflict resolution of '%@' failed: %@", self, doc.id, error);
+            dispatch_async(_otherDispatchQueue, ^{
+                // Conflict pulling a document -- the revision was added but app needs to resolve it
+                CBLLogInfo(Sync, @"%@: pulled conflicting version of '%@'", self, doc.id);
+                NSError* error;
+                if ([_config.database resolveConflictInDocument: doc.id
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: Make sure to catch exception here. If you plan to do it in a separate commit, please add TODO: here.

File: Objective-C/CBLConflict.h (hunk 3 above)
Comment (jayahariv) [inline Objective-C/CBLConflict.h]:
```suggestion
//  Copyright (c) 2019 Couchbase, Inc All rights reserved.
```

File: Objective-C/CBLConflict.m [hunk 11]
```diff
@@ -0,0 +1,38 @@
+//
+//  CBLConflict.m
+//  CouchbaseLite
+//
+//  Copyright (c) 2017 Couchbase, Inc All rights reserved.
```
Comment (jayahariv) [inline Objective-C/CBLConflict.m]:
```suggestion
//  Copyright (c) 2019 Couchbase, Inc All rights reserved.
```

File: Objective-C/CBLConflictResolution.h (hunk 4 above)
Comment (jayahariv) [inline Objective-C/CBLConflictResolution.h]:
```suggestion
//  Copyright (c) 2019 Couchbase, Inc All rights reserved.
```

File: Objective-C/CBLConflictResolution.m [hunk 12]
```diff
@@ -0,0 +1,52 @@
+//
+//  CBLConflictResolution.m
+//  CouchbaseLite
+//
+//  Copyright (c) 2018 Couchbase, Inc All rights reserved.
```
Comment (jayahariv) [inline Objective-C/CBLConflictResolution.m]:
```suggestion
//  Copyright (c) 2019 Couchbase, Inc All rights reserved.
```

File: Objective-C/CBLConflictResolver.h (hunk 5 above)
Comment (jayahariv) [inline Objective-C/CBLConflictResolver.h]:
```suggestion
//  Copyright (c) 2019 Couchbase, Inc All rights reserved.
```

File: Objective-C/CBLReplicator.mm (hunk 8 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: this can be used for any thing other than normal database serial queue right. so if we name only for conflict resolution, that might only be used for one purpose. Your thoughts?

NB: also how about `_concurrentQueue?`

File: Objective-C/CBLDatabase.mm (hunk 6 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: done

File: Objective-C/CBLDatabase.mm (hunk 7 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: done

File: Objective-C/CBLReplicator.mm (hunk 10 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: added TODO:

**Comment by jayahariv** [comment]
assigning back to you, since i made the changes, and review whenever you get time.

File: Objective-C/CBLDatabase.mm [hunk 13]
```diff
@@ -980,16 +980,39 @@ - (bool) resolveConflictInDocument: (NSString*)docID
         }
         
         conflictResolver = conflictResolver ?: [CBLConflictResolution default];
+        
+        CBLAssertNotNil(localDoc);
+        CBLAssertNotNil(remoteDoc);
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: It's possible localDoc or remoteDoc could be null.

File: Objective-C/CBLReplicator.mm (hunk 8 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: i named it `_concurrentQueue ` for now, do you have objections?

Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: updated name to `_conflictResolverQueue `, till we have other purpose for concurrent operations.

**Comment by jayahariv** [comment]
travis passed, but coveralls failed. unit tests will be added separately.

File: Objective-C/CBLReplicator.mm [hunk 14]
```diff
@@ -565,9 +565,12 @@ - (void) onDocsEnded: (NSArray<CBLReplicatedDocument*>*)docs pushing: (BOOL)push
                                                           error: &error]) {
                     [doc resetError]; // Reset the error as successfully resolving the conflicts
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: Reset the error here has no point as the ReplicatedDocument has already been notified with the conflict error.

File: Objective-C/CBLReplicator.mm [hunk 15]
```diff
@@ -565,9 +565,12 @@ - (void) onDocsEnded: (NSArray<CBLReplicatedDocument*>*)docs pushing: (BOOL)push
...
                 } else {
-                    // TODO: Handle the errors when the conflict happens, with the pull
                     CBLWarn(Sync, @"%@: Conflict resolution of '%@' failed: %@",
                             self, doc.id, error);
+                    id replication = [[CBLDocumentReplication alloc] initWithReplicator: self
+                                                                                 isPush: pushing
+                                                                              documents: @[doc]];
+                    [_docReplicationNotifier postChange: replication];
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: I'm not quite sure about the logic here as the ReplicatedDocument has already been notified with the conflict error. Now it will get notified again with another error.

I think we need to revise the whole logic around here. It could be that we will not notify ReplicatedDocument until the conflict gets resolved or has an error. Can you implement this logic and create a new issue to review the logic?

File: Objective-C/CBLDatabase.mm [hunk 16]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
-        CBLDocument* remoteDoc = [[CBLDocument alloc] initWithDatabase: self
-                                                            documentID: docID
-                                                        includeDeleted: YES
-                                                                 error: outError];
-        if (!remoteDoc || ![remoteDoc selectConflictingRevision])
-            return false;
+        Assert(localDoc == nil && remoteDoc == nil, @"Local and remote document shouldn't be empty \
+               at same time, when resolving conflict.");
+        localDoc = localDoc.isDeleted ? nil : localDoc;
+        remoteDoc = remoteDoc.isDeleted ? nil : remoteDoc;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: I think localDoc or remoteDoc here should be set to nil because the log messages in line 994 and the logic in line 1018 will not be correct. You can do this when creating CBLConflict directly.

File: Objective-C/CBLDatabase.mm [hunk 17]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
+        CBLConflict* conflict = [[CBLConflict alloc] initWithLocalDocument: localDoc
+                                                            remoteDocument: remoteDoc];
         
         // Resolve conflict:
-        CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
-                   docID, localDoc.revID, remoteDoc.revID);
+        CBLDocument* resolvedDoc;
+        @try {
+            CBLLogInfo(Sync, @"Resolving doc '%@' (mine=%@ and theirs=%@)",
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: change mine to local and theirs to remote to make it consistency with the naming in CBLConflict.

File: Objective-C/CBLDatabase.mm [hunk 18]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
+                       docID, localDoc.revID, remoteDoc.revID);
+            
+            resolvedDoc = [conflictResolver resolve: conflict];
+            
+            if (resolvedDoc.id != docID) {
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: resolvedDoc could be nil when returning a deleted doc.

File: Objective-C/CBLDatabase.mm [hunk 19]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
+                [NSException raise: NSInternalInconsistencyException
+                            format: @"Resolved docID '%@' is not matching with docID '%@'",
+                 resolvedDoc.id, docID];
+            }
+            
+            if (resolvedDoc.database != self) {
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: resolvedDoc could be nil so this check is not valid if resolved doc is nil.

File: Objective-C/CBLDatabase.mm [hunk 20]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
-                               mergedBody,
-                               mergedFlags,
-                               &c4err)
-        || !c4doc_save(rawDoc, 0, &c4err)) {
-        return convertError(c4err, outError);
+    CBL_LOCK(self) {
+        C4Transaction t(_c4db);
+        if (!t.begin())
+            return convertError(t.error(), outError);
+
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: If resolvedDoc is nil, we could try to match it to localDoc or remoteDoc only if one of them is a deletedDoc here.

File: Objective-C/CBLDatabase.mm [hunk 21]
```diff
@@ -950,102 +952,127 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
+        CBLStringBytes winningRevID = remoteDoc.revID;
+        CBLStringBytes losingRevID = localDoc.revID;
+        
+        alloc_slice mergedBody;
+        C4RevisionFlags mergedFlags = 0;
+        if (resolvedDoc != remoteDoc) {
+            // Unless the remote revision is being used as-is, we need a new revision:
+            mergedBody = [resolvedDoc encode: outError];
+            if (!mergedBody)
+                return false;
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: if resolvedDoc == nil, we don't need to call encode: and we shouldn't skip the check here too.

File: Objective-C/CBLReplicator.mm [hunk 22]
```diff
@@ -108,6 +110,9 @@ - (instancetype) initWithConfig: (CBLReplicatorConfiguration *)config {
         _progressLevel = kCBLProgressLevelBasic;
         _changeNotifier = [CBLChangeNotifier new];
         _docReplicationNotifier = [CBLChangeNotifier new];
+        
+        NSString* qName = $sprintf(@"DB-Replicator-Other <%@>", config.database.name);
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: What does the 'Other' mean?

File: Objective-C/CBLReplicator.mm (hunk 15 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: I can add a separate PR for whole error handling. keeping this same logic with a TODO comment saying to revert this.

File: Objective-C/CBLReplicator.mm (hunk 22 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: other process, outside transaction. I can rename this to, `Conflict resolver`

File: Objective-C/CBLDatabase.mm (hunk 18 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: Yea!! Seems like, I need to revisit whole PR again with nil handling.

File: Objective-C/CBLReplicator.mm (hunk 14 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: i am not changing this for now, will do this handling in separate PR.

File: Objective-C/CBLReplicator.mm (hunk 22 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: added `Conflict resolver <dbName>`

File: Objective-C/CBLDatabase.mm [hunk 23]
```diff
@@ -950,102 +952,138 @@ - (BOOL) saveDocument: (CBLDocument*)document
...
+        if (!t.begin())
+            return convertError(t.error(), outError);
+        
+        if (!resolvedDoc) {
+            if (!localDoc && remoteDoc)
+                resolvedDoc = remoteDoc; // local deleted, keep remote
+            
+            if (!remoteDoc && localDoc)
+                resolvedDoc = localDoc; // remote deleted, keep local
+        }
```
Comment (pasin) [inline Objective-C/CBLDatabase.mm]: localDoc and remoteDoc will never be nil here. I think the logic should be something like below (You may need to clean it up):

```
if (!resolvedDoc) {
   if (localDoc.isDeleted)
       resolved = localDoc
   else if (remoteDoc.isDeleted)
      resolved = remoteDoc
}
```

File: Objective-C/CBLDatabase.mm (hunk 21 above)
Comment (pasin) [reply Objective-C/CBLDatabase.mm]: I think the logic here might be something like this (You may need to clean it up):

```
if (!resolvedDoc) {
    mergedFlags |= kRevDeleted;
} else if (resolvedDoc != remoteDoc) {
    mergedBody = [resolvedDoc encode: outError];
    if (!mergedBody)
        return false;
    if (resolvedDoc.isDeleted)
        mergedFlags |= kRevDeleted;
}
```

Comment (pasin) [reply Objective-C/CBLDatabase.mm]: or 

```
if (resolvedDoc != remoteDoc) {
    if (resolvedDoc) {
        mergedBody = [resolvedDoc encode: outError];
        if (!mergedBody)
            return false;
    }

    BOOL isDeleted = !resolvedDoc || resolvedDoc.isDeleted
    if (isDeleted)
        mergedFlags |= kRevDeleted;
}
```

File: Objective-C/CBLDatabase.mm (hunk 23 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: we are assigning nil, if the localDoc or remoteDoc is deleted in the above function right? line 986, 987.

File: Objective-C/CBLDatabase.mm (hunk 16 above)
Comment (pasin) [reply Objective-C/CBLDatabase.mm]: I have a typo in the original comment. Can you review check it again?

File: Objective-C/CBLDatabase.mm (hunk 21 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]: is there a chance resolvedDoc is nil, even though the local and remote is not deleted? if any of them is deleted, then the resolvedDoc will not be nil. they will be populated by the logic above 

if that is the case, we will mark the remote as deleted by replacing the local contents right?

Comment (pasin) [reply Objective-C/CBLDatabase.mm]: > is there a chance resolvedDoc is nil, even though the local and remote is not deleted?

Yes. The API doesn't prevent at this moment.

> if any of them is deleted, then the resolvedDoc will not be nil. they will be populated by the logic above

This is correct.

> if that is the case, we will mark the remote as deleted by replacing the local contents right? which is wrong!

I couldn't follow this. Can you explain more?

File: Objective-C/CBLDatabase.mm (hunk 23 above)
Comment (pasin) [reply Objective-C/CBLDatabase.mm]: Yes

File: Objective-C/CBLDatabase.mm (hunk 21 above)
Comment (jayahariv) [reply Objective-C/CBLDatabase.mm]:
```
        alloc_slice mergedBody;
        C4RevisionFlags mergedFlags = 0;
        // resolvedDoc= nil
        // remote != nil
        // local != nil
        if (resolvedDoc != remoteDoc) {
            BOOL isDeleted = YES;
            if (resolvedDoc) {
                // Unless the remote revision is being used as-is, we need a new revision:
                mergedBody = [resolvedDoc encode: outError];
                if (!mergedBody)
                    return false;
                isDeleted = resolvedDoc.isDeleted;
            }
            
            if (isDeleted) // if resolved doc = nil, meaning the mergedFlags includes kRevDeleted
                mergedFlags |= kRevDeleted;
        }
        // mergedBody  = empty, since resolvedDoc = nil
        // winningRevId is remote.

//  remote will win, and body will be empty. 
```

Comment (pasin) [reply Objective-C/CBLDatabase.mm]: This is correct. Remote should always win. I think empty body is fine because the content of the deleted doc will be empty anyway. If LiteCore doesn't allow nil, we will need to create an empty body.

File: Objective-C/Internal/Replicator/CBLConflict+Internal.h [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#import "CBLConflict.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CBLConflict (Internal)
```
Comment (pasin) [inline Objective-C/Internal/Replicator/CBLConflict+Internal.h]: If declaring (Internal) category, don't you need to add implementation with (Internal) marked?

Comment (jayahariv) [reply Objective-C/Internal/Replicator/CBLConflict+Internal.h]: implementation functions are in the CBLConflict.m similar to the CBLDatabase, CBLReplicator

Comment (jayahariv) [reply Objective-C/Internal/Replicator/CBLConflict+Internal.h]: i think for the sake of consistency, i should rename that to `CBLConflict ()`

Comment (jayahariv) [inline Objective-C/Internal/Replicator/CBLConflict+Internal.h]:
```suggestion
@interface CBLConflict ()
```

Comment (pasin) [reply Objective-C/Internal/Replicator/CBLConflict+Internal.h]: Normally when using category, I remember that we need to do implementation with category name as well. Maybe it doesn't really need to do that.

File: Objective-C/CBLReplicator.mm (hunk 15 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: https://github.com/couchbase/couchbase-lite-ios/issues/2444
